### PR TITLE
feat(ci): add verification and force feature of docker images in release flow

### PR DESCRIPTION
**Pull request by bill-mahoney**
feat(ci): add verification and force feature of docker images in release flow.

**Would prefer to wait until after Hanoi releases to merge this.**

Signed-off-by: Bill Mahoney <bill.mahoney@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/edgexfoundry/ci-management/blob/master/.github/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Added labels
## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Issue Number:
https://github.com/edgexfoundry/cd-management/issues/40

## Sandbox Testing

Test Links :
[Docker Hub Link](https://hub.docker.com/r/edgexfoundry/docker-sample-service/tags)
Initial State: 
`sample-service 4.1.25 exists on dockerhub.`

First Build:
Push 4.1.26, no force flag -- Success
[Build Job](https://jenkins.edgexfoundry.org/job/edgexfoundry/job/cd-management/job/PR-88/12/console)
Output
`15:45:29  [edgeXReleaseDockerImage] Successfully published [2] images`

Second Build:
Push 4.1.26 again, no force flag -- Does not push:
[Build Job](https://jenkins.edgexfoundry.org/job/edgexfoundry/job/cd-management/job/PR-88/13/console)
Output:
`[edgeXReleaseDockerImage] Image exists at destination and option 'dockerForce' was not set in release.yaml file. [docker.io/edgexfoundry/docker-sample-service:4.1.26] did not release...`

Third Build:
Push 4.1.26 again, with dockerForce flag -- Success / Overwite
[Build Job](https://jenkins.edgexfoundry.org/job/edgexfoundry/job/cd-management/job/PR-88/14/console)
Output:
`[edgeXReleaseDockerImage] Successfully published [2] images`

Fourth Build:
Showing "Unstable" if docker image already exists and 'dockerForce' not used (added on 11/9/2020 from demo feedback)
[Build Job](https://jenkins.edgexfoundry.org/blue/organizations/jenkins/edgexfoundry%2Fcd-management/detail/PR-88/16/pipeline/43)
Output:
`[2020-11-09T23:12:02.193Z] WARNING: A Docker Image did not release`

## Are there any new imports or modules? If so, what are they used for and why?
No

### CODE REVIEW

File: vars/edgeXReleaseDockerImage.groovy [hunk 1]
```diff
@@ -144,4 +151,8 @@ def validate(releaseYaml) {
     if(!releaseYaml.version) {
         error("[edgeXReleaseDockerImage] Release yaml does not contain release 'version'. Example: v1.1.2")
     }
+}
+
+def imageExists(image) {
+    sh(script: "docker pull ${edgeXDocker.toImageStr(image)} > /dev/null && exit 1 || exit 0", returnStatus: true)
```
Comment (bill-mahoney) [inline vars/edgeXReleaseDockerImage.groovy:157]: docker pull returns 0 for successful pull. Switching the logic helps to satisfy the 'imageExists' wording (true = docker pull exit 0, false = docker pull exit > 0

File: src/test/groovy/edgeXReleaseDockerImageIntSpec.groovy [hunk 2]
```diff
@@ -186,5 +186,50 @@ public class EdgeXReleaseDockerImageIntSpec extends JenkinsPipelineSpecification
...
+                ]
+                
+                getPipelineMock('sh').call(script: "docker pull docker.io/edgexfoundry/docker-app-functions-sdk-go:v1.2.0 > /dev/null && exit 1 || exit 0", returnStatus: true) >> true
+            
+                edgeXReleaseDockerImage.getBinding().setVariable('env', environmentVariables)
+                edgex.getBinding().setVariable('env', environmentVariables)
+                validReleaseYaml.dockerForce = true
+                
+                edgeXReleaseDockerImage.getBinding().setVariable('env', environmentVariables)
+                edgex.getBinding().setVariable('env', environmentVariables)
```
Comment (soda480) [inline src/test/groovy/edgeXReleaseDockerImageIntSpec.groovy:224]: lines 223 & 224 are not necessary - repeated from lines 219 & 220

**Review by soda480 — APPROVED** [review]
Looks good, I just have a minor recommendation on some repeated code you had in one of your unit tests.

**Review by ernestojeda — DISMISSED** [review]
👍

**Review by cjoyv — DISMISSED** [review]
Looks good

File: src/test/groovy/edgeXReleaseDockerImageIntSpec.groovy (hunk 2 above)
Comment (bill-mahoney) [reply src/test/groovy/edgeXReleaseDockerImageIntSpec.groovy:222]: Thanks, resolved

**Review by soda480 — APPROVED** [review]
LGTM

**Pull request merged**